scandium-core/src/main/java/org/eclipse/californium/scandium/dtls/ConnectionIdExtension.java

```diff
@@ -0,0 +1,129 @@
+/*******************************************************************************
+ * Copyright (c) 2019 Bosch Software Innovations GmbH and others.
+ * 
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v1.0
+ * and Eclipse Distribution License v1.0 which accompany this distribution.
+ * 
+ * The Eclipse Public License is available at
+ *    http://www.eclipse.org/legal/epl-v10.html
+ * and the Eclipse Distribution License is available at
+ *    http://www.eclipse.org/org/documents/edl-v10.html.
+ * 
+ * Contributors:
+ *    Bosch Software Innovations - initial creation
+ ******************************************************************************/
+package org.eclipse.californium.scandium.dtls;
+
+import java.net.InetSocketAddress;
+
+import org.eclipse.californium.elements.util.DatagramWriter;
+import org.eclipse.californium.scandium.dtls.AlertMessage.AlertDescription;
+import org.eclipse.californium.scandium.dtls.AlertMessage.AlertLevel;
+
+/**
+ * Conveys information specified by the <em>connection id</em> DTLS extension.
+ * <p>
+ * See <a href=
+ * "https://datatracker.ietf.org/doc/draft-ietf-tls-dtls-connection-id">draft-ietf-tls-dtls-connection-id</a>
+ * for additional details.
+ *
+ */
+public final class ConnectionIdExtension extends HelloExtension {
+
+	/**
+	 * Number of bits for the encoded length of the connection id in the
+	 * extension.
+	 */
+	private static final int CID_FIELD_LENGTH_BITS = 8;
+
+	/**
+	 * Connection id to negotiate.
+	 */
+	private ConnectionId id;
+
+	/**
+	 * Create connection id extension.
+	 * 
+	 * @param id connection id
+	 */
+	private ConnectionIdExtension(ConnectionId id) {
+		super(ExtensionType.CONNECTION_ID);
+		this.id = id;
+	}
+
+	/**
+	 * Get connection id.
+	 * 
+	 * @return connection id
+	 */
+	public ConnectionId getConnectionId() {
+		return id;
+	}
+
+	@Override
+	public int getLength() {
+		// 2 bytes indicating extension type, 2 bytes overall length,
+		// 1 byte cid length + cid
+		return 2 + 2 + 1 + id.length();
+	}
+
+	@Override
+	protected void addExtensionData(final DatagramWriter writer) {
+		int length = id.length();
+		writer.write(1 + length, LENGTH_BITS);
+		writer.write(length, CID_FIELD_LENGTH_BITS);
+		writer.writeBytes(id.getBytes());
```

Comment:
This is an arbitrary way to encode CID ? 
I mean there is nothing which say that first byte is the length and next byte are CID ?
We could have used fixed size CID right ?

Reply:
> I mean there is nothing which say that first byte is the length and next byte are CID ?

The draft encodes the cid different in the hello extension and the new record type.
In the extension with leading uint8 length, and opaque bytes, in the new record type only the opaque types. That's specified by the draft.

Anyway the draft allows you to be more flexible as my current implementation. The draft specifies the cid similar to the coap-token. If one peer send the cid in the hello extension, then the other, if agreed, sent that cid in the records of the next epoch. If a peer uses different sized cids, it must do it in a proper way, so that the peer will be able to read them back.

Reply:
Ok I missed the meaning of https://tools.ietf.org/html/rfc5246#section-4.3 notation